### ISSUE TITLE
IUrlHelper does not seems to work when used in `IBackgroundTask`

### ISSUE THREAD

**Issue by MikeAlhayek** [OrchardCMS/OrchardCore#14086]
### Describe the bug
I have a background task that runs every 1 hours a checks for new published content items and sends out notifications. 

I want to be able to send out the `display_url` of each content item so that the user can click on the link from their email to view the correct content item.

I used Liquid to create the email body. In the email I do something like this `{{ Model.ContentItem | display_url }}`. But, for some reason the value is always nil. So the url isn't being generated for some reason. The following line https://github.com/OrchardCMS/OrchardCore/blob/83ada8fbb11e208caab7d3111b33deb056009d1a/src/OrchardCore.Modules/OrchardCore.Contents/Liquid/DisplayUrlFilter.cs#L53 seems to always return null even when `routeValues` has the correct values.

I am assuming the issue is because the request is coming from the background task and there is not URL in the HttpRequest but not positive. 

Is there a workaround for this?

To reproduce this issue you can add this Background task
```
[BackgroundTask(
    Title = "Testing",
    Schedule = "* * * * *",
    Description = "Testing")]
public class TestingBackgroundTask : IBackgroundTask
{
    public async Task DoWorkAsync(IServiceProvider serviceProvider, CancellationToken cancellationToken)
    {
        var session = serviceProvider.GetRequiredService<ISession>();
        var liquidTemplateManager = serviceProvider.GetRequiredService<ILiquidTemplateManager>();
        var htmlEncoder = serviceProvider.GetRequiredService<HtmlEncoder>();

        var contentItem = await session.Query<ContentItem, ContentItemIndex>(x => x.Published).FirstOrDefaultAsync();

        var message = "{{ ContentItem | display_url }}";

        var result = await liquidTemplateManager.RenderHtmlContentAsync(message, htmlEncoder, null,
            new Dictionary<string, FluidValue>()
            {
                ["ContentItem"] = new ObjectValue(contentItem),
            });

        using var writer = new StringWriter();
        result.WriteTo(writer, htmlEncoder);

        var finalString = writer.ToString();
    }
}
```

Then register it like this
```
services.AddSingleton<IBackgroundTask, TestingBackgroundTask>();
```
Put a break point and step into `DisplayUrlFilter`

**Comment by MikeAlhayek** [comment]
@jtkech from what I can tell is the `IRouteCollection` has no router to generate the virtual path. Did we change how the router is registered in OC?

**Comment by jtkech** [comment]
Yes, when rendering a liquid template in the background we create on the fly a `ViewContext` and `ActionContext` but the `Routers` only contains an empty `RouteCollection`.

So we can generate simple urls but no those based on routes that each modules configure in their startups when the tenant pipeline is built. I thought about it but we would need to do in the background (only once per tenant) what the `ModularTenantRouterMiddleware` does (on the first request) by calling `startup.Configure()` for each feature.

https://github.com/OrchardCMS/OrchardCore/blob/83ada8fbb11e208caab7d3111b33deb056009d1a/src/OrchardCore/OrchardCore/Modules/ModularTenantRouterMiddleware.cs#L121-L127

Maybe something crazy to try, but not sure it is recommended.

---

If your content item has an `AutoroutePart`, I tried the following that seems to work.

    var message = "{{ '/' | append: ContentItem.Content.AutoroutePart.Path | absolute_url }}";

To make it fully working for a given tenant you need to set its `Hostname` property (the one we see when editing a tenant), in the background by default we set it to `localhost` without any port.

**Comment by MikeAlhayek** [comment]
@jtkech calling `var app = serviceProvider.GetService<IApplicationBuilder>();` from the background task returns null. so even if we wan to call `UseRouting()...` we can't. 

>> Maybe something crazy to try, but not sure it is recommended.
Just curious why is it crazy and not recommended? Shouldn't the background task be able to generate URLs for the app correctly? 

>> var message = "{{ '/' | append: ContentItem.Content.AutoroutePart.Path | absolute_url }}";

`display_url` has better logic as you know since the app could change the URL even when not using `AutoroutePart`.

**Comment by MikeAlhayek** [comment]
Also, I noticed this while looking at the code
https://github.com/OrchardCMS/OrchardCore/blob/83ada8fbb11e208caab7d3111b33deb056009d1a/src/OrchardCore/OrchardCore.Abstractions/BackgroundTasks/ShellContextExtensions.cs#L31

Shouldn't the schema be set differently based on the current tenant?

**Comment by jtkech** [comment]
> calling var app = serviceProvider.GetService<IApplicationBuilder>(); from the background task returns null.

Yes for sure, it was just a link to code that call the module startup configure methods.

> Shouldn't the background task be able to generate URLs for the app correctly?

In a regular aspnet app, you can't do that in a `HostedService`, we already can do much more in our background tasks.

Maybe something possible but not so obvious, would need more time to see how we could populate a custom endpoint data source that could be then used by the aspnet helpers.

**Comment by jtkech** [comment]
> Shouldn't the schema be set differently based on the current tenant?

Normally that's okay, it is part of the shellSettings of the shellContext that is built in the same way, not the same as the pipeline that is built by calling startup `Configure()`, not `ConfigureServices()`.

**Comment by MikeAlhayek** [comment]
@jtkech yes I fixed the typo in my previous comment. At glance, hard coding `https` looked weird since there may be clients that don't use `https`. But you said it's okay. so it should not impact any generated absolute urls.

Just need to find a way to register the router so we are able to generate URL's from the background task.

**Comment by sebastienros** [comment]
> we would need to do in the background (only once per tenant) what the `ModularTenantRouterMiddleware` does (on the first request)

Or should we wait for a request to be sent, otherwise tenants would not be loaded lazily.
-> Are background tasks supposed to work if a tenant is not loaded (lazily).

We could also have a custom mode per tenant that would allow background tasks to run (or not being lazily loaded). So a tenant could either be delay-loaded (default) or eager-loaded (with just DI and background tasks maybe?).

**Comment by jtkech** [comment]
@sebastienros 

We already have a `BackgroundServiceOptions.ShellWarmup`, if true we don't wait that the shell pipeline is built to run background tasks, in that case a shell is eager-loaded and bluit but still no pipeline.

Anyway, even if the pipeline is built, in a background task we bluid a default HttpContext, when rendering a shape we also build on the fly an ActionContext and a ViewContext.

But for example the `ActionContext.RouteData.Routers` are not populated as it is done in a regular HttpRequest, we may need to retrieve endpoint data sources, I will take a look on how it is done, maybe in middlewares / filters, I don't know, and that are not executed in the background.

**Comment by jtkech** [comment]
@sebastienros @MikeAlhayek 

Okay I made it working, it was already possible to generate some urls but not those based on route values. Because of internal classes it's not so easy to repro what e.g. the endpoint middlewares do.

But crazy, I was able to run the shell pipeline in the `ModularBackgroundService`, to make it very fast, in one of our middleware (I will do a dedicated one) I shortcut the pipeline if we are in the background (we have an HttpContext item for that), but this still let e.g. the endpoint middlewares do their jobs.

Doing this made route generations working in the background.

I will suggest a PR.
